### PR TITLE
Add count of rentals for items and customers

### PR DESCRIPTION
**Pull request by skjerns**
I've added a functionality that counts how often an item has been rented, or how many rentals a customer has. it's loading lazily, so not slowing down the UI :)

It's added to the `keyValueStore["currentDoc"]["number_of_rentals"]` - that probably means that it's also written to the db. However, as there is no change no new shard will be created, right, and anyway only if save is clicked? Else let me know how I can display it without it being automatically saved to the doc, could not find out how to without creating a new Input field like `InputLabel` or so, so without bindings. Should I do it with a new `Input` Element that just displays things without binding?

For items
image

For customers
image

### CODE REVIEW

File: Frontend/src/components/Database/Database.js [hunk 1]
```diff
@@ -109,6 +109,17 @@ class Database {
     }
   }
 
+  async countDocs(selectors) {
```
Comment (daniel17903) [inline Frontend/src/components/Database/Database.js:112]: since you return a promise without awaiting it this function does not need to be async

**Comment by daniel17903** [comment]
Nice idea :) Yes, it is written to the database when "save" is clicked (see https://github.com/leih-lokal/LeihLokalVerwaltung/blob/689e1bd4043ab03a0743e348861ad33202255b1b/Frontend/src/components/TableEditors/Items/ItemPopupFormular.svelte#L249).

Wouldn't it be better to display the count in the table or in a separate tab for statistics? I think it might not be so good to display it as an input in the formular since it's not something the user can / should change. If we want to show more statistics later, maybe even some graphs, a separate tab for statistics (next to "Kunden" "Gegenstände" "Leihvorgänge") might be good?

**Comment by skjerns** [comment]
> Wouldn't it be better to display the count in the table or in a separate tab for statistics? I think it might not be so good to display it as an input in the formular since it's not something the user can / should change. If we want to show more statistics later, maybe even some graphs, a separate tab for statistics (next to "Kunden" "Gegenstände" "Leihvorgänge") might be good?

This is a good idea, however, that would but quite a load on the database, right, to query it individually for each item. We could however only query it for the currently displayed items, and/or find a way to only have one run through the db und sum up for the items, and keep that in memory? I don't really know how much of a strain this is, and what would be the most feasable solution. what do you suggest?

I could imagine having a separate tab for statistics! I wanted to do that anyway at some point :) but let's discuss soon what's the most urgent on our list, i guess statistics is rather a nice-to-have

**Comment by daniel17903** [comment]
> > Wouldn't it be better to display the count in the table or in a separate tab for statistics?
> 
> This is a good idea, however, that would but quite a load on the database, to query it individually for each item? We could however only query it for the currently displayed items, and/or find a way to only have one run through the db und sum up for the rentals-per-item, and keep that in local storage, or similar things? I don't really know how much of a strain this is, and what would be the most feasable solution. what do you suggest?
> 
> > If we want to show more statistics later, maybe even some graphs, a separate tab for statistics (next to "Kunden" "Gegenstände" "Leihvorgänge") might be good?
> > I could imagine having a separate tab for statistics! I wanted to do that anyway at some point :) but let's discuss soon what's the most urgent on our list, i guess statistics is rather a nice-to-have.
> 
> > I think it might not be so good to display it as an input in the formular since it's not something the user can / should change.
> 
> Mhh, I do agree that it's not an input, however it could be nice to link this to the individual pages, and not only in a summary tab.

I suggest we implement one of these solutions:

1. We could query the count for each item on the current page in the background and display it in the table. This could be implemented like the [cellBackgroundColorsFunction](https://github.com/leih-lokal/LeihLokalVerwaltung/blob/6dc2d53d957c962083e96db625c10e8804338aa4/Frontend/src/components/TableEditors/TableEditorConfig.js#L64) which is also executed asynchronously for each row on the current page. I think this would be the least effort but might decrease performance a bit, e.g. if the user already performs another action while the background tasks are still running. 
2. We could implement a python script that runs periodically, queries the counts for all items and stores them in the database. Then in the frontend we could just display the count. If we add more statistics later we might need this anyway because creating statistics might take too long to do it in the frontend.

**Comment by skjerns** [comment]
> > > Wouldn't it be better to display the count in the table or in a separate tab for statistics?
> > 
> > 
> > This is a good idea, however, that would but quite a load on the database, to query it individually for each item? We could however only query it for the currently displayed items, and/or find a way to only have one run through the db und sum up for the rentals-per-item, and keep that in local storage, or similar things? I don't really know how much of a strain this is, and what would be the most feasable solution. what do you suggest?
> > > If we want to show more statistics later, maybe even some graphs, a separate tab for statistics (next to "Kunden" "Gegenstände" "Leihvorgänge") might be good?
> > > I could imagine having a separate tab for statistics! I wanted to do that anyway at some point :) but let's discuss soon what's the most urgent on our list, i guess statistics is rather a nice-to-have.
> > 
> > 
> > > I think it might not be so good to display it as an input in the formular since it's not something the user can / should change.
> > 
> > 
> > Mhh, I do agree that it's not an input, however it could be nice to link this to the individual pages, and not only in a summary tab.
> 
> I suggest we implement one of these solutions:
> 
> 1. We could query the count for each item on the current page in the background and display it in the table. This could be implemented like the [cellBackgroundColorsFunction](https://github.com/leih-lokal/LeihLokalVerwaltung/blob/6dc2d53d957c962083e96db625c10e8804338aa4/Frontend/src/components/TableEditors/TableEditorConfig.js#L64) which is also executed asynchronously for each row on the current page. I think this would be the least effort but might decrease performance a bit, e.g. if the user already performs another action while the background tasks are still running.
> 2. We could implement a python script that runs periodically, queries the counts for all items and stores them in the database. Then anywhere in the frontend we could just display the count. If we add more statistics later we might need this anyway because creating statistics might take too long to do it in the frontend.

1) I'd vote for this solution. It should also not be too straining if it's only for the currently displayed items, right? However, it would not be possible to do sorting...

2) that would probably make it a bit harder to port it to other locations, as they have to be knowledable enough to setup a python environment? I'd prefer to keep things without a backend, and just let the statistics only load on demand (even if it takes a few seconds to load, that should not infer with the general performance of the app, unless triggered to run).

**Comment by skjerns** [comment]
How do we want to handle this PR? 
Do you want to merge it for now, or should we add another category "Statistik/Information/[..]" within the popup (I could imagine other information placed there: Which item was rented last, last interaction, etc etc)

**Comment by daniel17903** [comment]
> How do we want to handle this PR?
> Do you want to merge it for now, or should we add another category "Statistik/Information/[..]" within the popup where we place this as new Input method that only displays/a label (I could imagine other information placed there: Which item was rented last, last interaction, etc etc) or do you want to discard it altogether?

I would prefer to update and then merge this PR, I'm sure we can use parts of your current implementation.

I still don't like displaying statistics in the popup formular that much since it's original purpose was editing items and it already is quite large. Moving it to a separate section as you suggested would be better. I would still prefer to show statistics in the table or in a separate tab. 

But since you implement this you can decide. Showing statistics in the popup formular might still be better than not showing statistics at all.

**Pull request closed without merging**